contributors/design-proposals/storage/csi-snapshot.md

```diff
@@ -0,0 +1,395 @@
+
+#### The `VolumeSnapshot` Object
+
+```GO
+
+// The volume snapshot object accessible to the user. Upon successful creation of the actual
+// snapshot by the volume provider it is bound to the corresponding VolumeSnapshotContent through
+// the VolumeSnapshotSpec
+type VolumeSnapshot struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec represents the desired state of the snapshot
+	// +optional
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status represents the latest observer state of the snapshot
+	// +optional
+	Status VolumeSnapshotStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Items           []VolumeSnapshot `json:"items"`
+}
+
+// VolumeSnapshotSpec is the desired state of the volume snapshot
+type VolumeSnapshotSpec struct {
+        // PersistentVolumeClaimName is the name of the PVC being snapshotted
+        // +optional
+        PersistentVolumeClaimName string `json:"persistentVolumeClaimName" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+        // SnapshotDataName binds the VolumeSnapshot object with the VolumeSnapshotContent
+        // +optional
+        SnapshotDataName string `json:"snapshotDataName" protobuf:"bytes,2,opt,name=snapshotDataName"`
+
+        // Name of the SnapshotClass required by the volume snapshot.
+        // +optional
+        SnapshotClassName string `json:"storageClassName" protobuf:"bytes,3,opt,name=storageClassName"`
+}
+
+type VolumeSnapshotCreated struct {
+        // The time the snapshot was successfully created
+        // +optional
+	CreatedAt metav1.Time `json:"createdAt" protobuf:"bytes,1,opt,name=createdAt"`
+}
+
+type VolumeSnapshotAvailable struct {
+        // The time the snapshot was successfully created and available for use
+        // +optional
+        AvailableAt metav1.Time `json:"availableAt" protobuf:"bytes,1,opt,name=availableAt"`
+}
+
+type VolumeSnapshotError struct {
+        // A brief CamelCase string indicating details about why the snapshot is in error state.
+        // +optional
+        Reason string
+        // A human-readable message indicating details about why the snapshot is in error state.
+        // +optional
+        Message string
+        // The time the error occurred during the snapshot creation (or uploading) process
+        // +optional
+        FailedAt metav1.Time `json:"failedAt" protobuf:"bytes,1,opt,name=failedAt"`
+}
+
+type VolumeSnapshotStatus struct {
+	// VolumeSnapshotCreated indicates whether the snapshot was successfully created.
+	// If the timestamp CreateAt is set, it means the snapshot was created;
+	// Otherwise the snapshot was not created.
+        // +optional
+        Created VolumeSnapshotCreated
```

Comment:
Since it's optional, should it be a pointer?

```go
if snap.Status.Created == nil  {
}
```

vs.

```go
if snap.Status.Created.CreatedAt.IsZero() {
}
```

Reply:
Why do we need this to be a struct at all? The "IsZero()" test would work with the plain timestamp too and should be sufficient to figure out if the snapshot exists and since when.

Reply:
Since we only have timestamp in the struct now, it is possible to get rid of the struct.  A struct would allow us to add more parameters to it later if needed.

Reply:
I think that's a good suggestion. We can make it like this

```
Type VolumeSnapshotStatus struct {
    CreatedAt *metav1.Time
    AvailableAt *metav1.Time
    FailedAt *metav1.Time
    Reason string
    Message string
}
```